src/cluster/serializer/binary-serialization-format/Serializer.cc

```diff
@@ -0,0 +1,191 @@
+#include "zeek/cluster/serializer/binary-serialization-format/Serializer.h"
+
+#include <cstdint>
+
+#include "zeek/ID.h"
+#include "zeek/Reporter.h"
+#include "zeek/SerializationFormat.h"
+#include "zeek/Type.h"
+#include "zeek/Val.h"
+#include "zeek/cluster/Serializer.h"
+#include "zeek/logging/Types.h"
+#include "zeek/threading/SerialTypes.h"
+
+using namespace zeek::cluster;
+
+// #define SERIALIZER_DEBUG(...) fprintf(stderr, __VA_ARGS__)
+#define SERIALIZER_DEBUG(...)                                                                                          \
+    do {                                                                                                               \
+    } while ( 0 )
+
+bool detail::BinarySerializationFormatLogSerializer::SerializeLogWrite(byte_buffer& buf,
+                                                                       const logging::detail::LogWriteHeader& header,
+                                                                       zeek::Span<logging::detail::LogRecord> records) {
+    zeek::detail::BinarySerializationFormat fmt;
+
+    SERIALIZER_DEBUG("Serializing stream=%s writer=%s filter=%s path=%s num_fields=%zu num_records=%zu\n", stream_id,
+                     writer_id, header.filter_name.c_str(), header.path.c_str(), header.fields.size(), records.size());
+
+    fmt.StartWrite();
+
+    // Header: stream_name, writer_id, filter_name, path, num_fields, schema fields
+    bool success = true;
+    success &= fmt.Write(header.stream_name, "stream_id");
+    success &= fmt.Write(header.writer_name, "writer_id");
+    success &= fmt.Write(header.filter_name, "filter_name");
+    success &= fmt.Write(header.path, "path");
+    success &= fmt.Write(static_cast<uint32_t>(header.fields.size()), "num_fields");
+    for ( const auto& f : header.fields )
+        success &= f.Write(&fmt);
+
+    success &= fmt.Write(static_cast<uint32_t>(records.size()), "num_records");
+
+    if ( ! success ) {
+        reporter->Error("Failed to remotely log stream %s: header serialization failed", header.stream_name.c_str());
+        return false;
+    }
+
+    // Write out the payload.
+    for ( const auto& rec : records ) {
+        for ( size_t i = 0; i < rec.size(); ++i ) {
+            if ( ! rec[i].Write(&fmt) ) {
+                reporter->Error("Failed to remotely log stream %s: field %zu serialization failed",
+                                header.stream_name.c_str(), i);
+                return false;
+            }
+        }
+    }
+
+    char* data;
+    uint32_t len;
+    len = fmt.EndWrite(&data);
+
+    // Copy result into buffer. Would be nice to serialize directly into the buffer,
+    // but the SerializationFormat doesn't really allow for that.
```

Comment:
Broker also copies, but into a `std::string`: - within cluster `vector<std::byte>` is used as a container rather than `std::string`. If we we keep with SerializationFormat rather than using something else altogether might be worth changing this to serialize into an appropriate container.

https://github.com/zeek/zeek/blob/2c52e533b222357b5180acbb7407431dce2939ee/src/broker/Manager.cc#L693-L695